text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,880 @@
+cargo update -p anstyle --precise 1.0.2
+    Updating crates.io index
+ Downgrading anstyle v1.0.4 -> v1.0.2
+$ cargo +1.64.0 check
+  Downloaded anstyle v1.0.2
+  Downloaded 1 crate (14.0 KB) in 0.60s
+   Compiling rustix v0.38.23
+    Checking bitflags v2.4.1
+    Checking linux-raw-sys v0.4.11
+    Checking utf8parse v0.2.1
+    Checking anstyle v1.0.2
+    Checking colorchoice v1.0.0
+    Checking anstyle-query v1.0.0
+    Checking clap_lex v0.5.0
+    Checking strsim v0.10.0
+    Checking anstyle-parse v0.2.1
+    Checking is-terminal v0.4.9
+    Checking anstream v0.3.2
+    Checking clap_builder v4.3.24
+    Checking clap v4.3.24
+    Checking msrv-resolver v0.1.0 (/home/epage/src/personal/dump/msrv-resolver)
+    Finished dev [unoptimized + debuginfo] target(s) in 2.96s
+```
+Success! Mixed with many tears and less hair.
+
+How wide spread is this?  Take this with a grain of salt but based on crates.io user agents:
+
+| Common MSRVs        | % Compatible Requests |
+|--------------------:|:----------------------|
+| N (`1.73.0`)        | 47.432%               |
+| N-2 (`1.71.0`)      | 74.003%               |
+| ~6 mo (`1.69.0`)    | 93.272%               |
+| ~1 year (`1.65.0`)  | 98.766%               |
+| Debian (`1.63.0`)   | 99.106%               |
+| ~2 years (`1.56.0`) | 99.949%               |
+
+*([source](https://rust-lang.zulipchat.com/#narrow/stream/318791-t-crates-io/topic/cargo.20version.20usage/near/401440149))*
+
+People have tried to reduce the pain from MSRV with its own costs:
+- Treating it as a breaking change:
+  - This leads to extra churn in the ecosystem when a fraction of users are likely going to benefit
+  - We have the precedence elsewhere in the Rust ecosystem for build and runtime system requirement changes not being breaking, like when rustc requires newer versions of glibc, Android NDK, etc.
+- Adding upper limits to version requirements:
+  - This fractures the ecosystem by making packages incompatible with each other and the Cargo team [discourages doing this](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#multiple-requirements)
+- Avoiding dependencies, re-implementing it themselves at the cost of their time and the risk for bugs, especially if `unsafe` is involved
+- Ensuring dependencies have a more inclusive MSRV policy then themselves
+  - This has lead to long arguments in the ecosystem over what is the right
+    policy for updating a minimum-supported Rust version (MSRV),
+    wearing on all
+    (e.g.
+    [libc](https://github.com/rust-lang/libs-team/issues/72),
+    [time](https://github.com/time-rs/time/discussions/535)
+    )
+
+The sooner we improve the status quo, the better, as it can take years for
+these changes to percolate out to those exclusively developing with an older
+Rust version (in contrast with the example above).
+This delay can be reduced somewhat if a newer toolchain can be used for
+development version without upgrading the MSRV.
+
+## Workflows
```

Comment:
@LukeMathWalker 

> I wanted to leave targeted feedback on the RFC paragraphs, but I'm struggling since it isn't structured as I would expect it to. You've identified workflows, but even before that I would expect to see _user personas_. Different user personas have different workflows attached, and I think the RFC overindexes on library maintainers ("application developers" only appears once in the entire proposal).
> 
> I raise this concern since I believe there's a significant difference between the needs of a _library maintainer_ and those of an _application developer_.
> 
> As an application developer, for example, I would be surprised to find out that the default resolver scheme is not `maximum` or, at the very least, `rustc`. My primary concern, as expressed before by @jonhoo, is to maximise the likelihood of running a version of my dependencies that doesn't contain known bugs or security vulnerabilities. Targeting a version of `rustc` that's not `latest` should be a deliberate choice, executed in full awareness of the consequences.
> 
> I expect library maintainers to be, on average, more sophisticated Rust users than application developers. The latter is also the largest group of users. Defaults should be picked by putting the needs of application developers at the forefront, rather than those of library maintainers. I realize this might be controversial, but I haven't seen this opinion stated clearly in the tens of comments I had a chance to read so far.

*(moved from [here](https://github.com/rust-lang/rfcs/pull/3537#issuecomment-1902131401) to avoid interleaved conversations)*

Reply:
> Different user personas have different workflows attached, and I think the RFC overindexes on library maintainers ("application developers" only appears once in the entire proposal).

Part of the reason that term only appears once is the focus hasn't been so much in those terms.  

imo the "Latest MSRV" workflows are agnostic of app vs lib.

The "Extended MSRV" workflow acknowledges a lot of different user groups and I interpret most of those as talking about app developers.  Only the last bullet item ("Library and tool maintainers catering to the above use cases") has libraries and still it is shared with app developers.

> As an application developer, for example, I would be surprised to find out that the default resolver scheme is not maximum or, at the very least, rustc. 

When I enumerated "Extended MSRV with latest dependencies", I only found it applicable to "Library and tool maintainers catering to the above use cases".  Is that you?  If not, that means there is a gap and I would love to better understand your situation, including
- What type of application?
- What is your MSRV policy?
- Why do you have that MSRV policy?
- How do you verify your MSRV?
- Is the application meant to be installable with `cargo install`?
  - This is important because Cargo expects that `Cargo.lock` is compatible with your MSRV today and tells to users to pass `--locked` on MSRV failures.  This also gets into the definition of the word "support" within MSRV on what all interactions with a project an MSRV applies to.

Reply:
> Part of the reason that term only appears once is the focus hasn't been so much in those terms.

I understand that's not the framing being used. Take it as a feedback point: I think reframing the excellent work done on workflows around the different types of user intent would help focus the conversation and make the RFC more approachable/easier to read.  
Even talking about "minimum supported Rust version" feels weird in the context of application development. It's not about _supporting_ a certain version of Rust, it's more about _using_ a certain version of Rust (or not caring about it).

>  Is that you?

The vast majority of Rust gigs I have been at operated under this framework:

- We are building an application that won't be distributed to users. It'll be running on a server/Cloud-platform of some kind.
- We care about our applications' performance, correctness and security above everything else. Consequently, we should always be using the latest versions of our dependencies.
- We don't have an MSRV. There might be a `rust-toolchain.toml` file in the repository to ensure that all developers are using the same version that's tested in CI, but the version is bumped up whenever someone remembers to do it or if there is a need (e.g. a dependency is using features from a newer version of Rust).

This feels closer to "Latest Rust as the MSRV", using the wording of the RFC.  
For this kind of workflow, `maximum` if the resolver strategy that makes most sense to me. I'd go as far as saying that I'd love if `rustup` auto-updated the `stable` toolchain for me when it detects I need something newer.

Reply:
>  Take it as a feedback point: I think reframing the excellent work done on workflows around the different types of user intent would help focus the conversation and make the RFC more approachable/easier to read.

I feel I'm missing something as I'm not seeing how this would help.  Could you elaborate more on what you envisioning and why it would help?  For me, I'm particularly looking to understand the value add it helps for the conversations vs the time sink it is to go and write up.  The current 4 workflows are a collapsing down of what I was originally looking at doing which was overwhelming enough that the only alternative is to not do this RFC.  So I'm coming from the perspective of "I've was thinking of something that vaguely sounds like what you're asking and it was untenable".  What your asking might be something completely different and thats why I'm wanting to better understand.  This might also be useful to do more directly.  I have the standing invite that people can come to cargo office hours to discuss this and to reach out to me directly to schedule if those times don't work.

> Even talking about "minimum supported Rust version" feels weird in application development. It's not about supporting a certain version of Rust, it's more about using a certain version of Rust (or not caring about it), under the (very common?) assumption that nobody but your own engineers will be building from source.

This is called out in the "Extended MSRV" workflow that sometimes they only support on version and they'll be setting it in `rust-toolchain.toml` instead of `package.rust-version`.

>  The vast majority of Rust gigs I have been at operated under this framework:

Thanks for the explanation!

For your use case, the RFC moves the message earlier (`update`, rather than the later `check`) and softens it (message, rather than error).  There are still errors if the MSRV is unset and uses incompatible features but that will be reduced over time as the RFC encourages MSRV to be set in more cases via the `"auto"` value.  For cases like yours where `rust-toolchain.toml` is used, we can't offer `--update-rust-version` (since that will only be for the manifest), we can at least suggest upgrading your toolchain instead.

While you might not be needing all of this hand holding, I hope the value of this hand holding is understandable within one of the prioritized workflow.  My expectation is that this won't get in the way of following "Latest MSRV" but will provide a friendlier experience.

As the RFC points out, there are other application developers in the "Extended MSRV" workflow that those changes for friendliness make a big difference for making their experience workable.

>  I'd go as far as saying that I'd love if rustup auto-updated the stable toolchain for me when it detects I need something newer.

With `toml_edit`, this is doable!  I recommend opening an issue for a command or flag to update a toolchain file.

Unsure how I feel about adding a future possibility of `--update-rust-version` modifying toolchain files.  That would require breaking abstractions and require cargo to have a lot more context from the rustup invocation.